pkg/archive/archive.go

```diff
@@ -225,6 +226,91 @@ func CompressStream(dest io.Writer, compression Compression) (io.WriteCloser, er
 	}
 }
 
+// TarModifierFunc is a function that can be passed to ReplaceFileTarWrapper to
+// define a modification step for a single path
+type TarModifierFunc func(path string, header *tar.Header, content io.Reader) (*tar.Header, []byte, error)
+
+// ReplaceFileTarWrapper converts inputTarStream to a new tar stream
+// while replacing a single file called header.Name with new contents.
+// If the file with header.Name does not exist it is added to the tar stream.
+// TODO: make this into a generic tar conversion function with walkFn argument
+func ReplaceFileTarWrapper(inputTarStream io.ReadCloser, mods map[string]TarModifierFunc) io.ReadCloser {
+	pipeReader, pipeWriter := io.Pipe()
+
+	modKeys := make([]string, 0, len(mods))
+	for key := range mods {
+		modKeys = append(modKeys, key)
+	}
+	sort.Strings(modKeys)
+
+	go func() {
+		tarReader := tar.NewReader(inputTarStream)
+		tarWriter := tar.NewWriter(pipeWriter)
+
+		defer inputTarStream.Close()
+
+	loop0:
+		for {
+			hdr, err := tarReader.Next()
+			for len(modKeys) > 0 && (err == io.EOF || err == nil && hdr.Name >= modKeys[0]) {
+				var h *tar.Header
+				var rdr io.Reader
+				if hdr != nil && hdr.Name == modKeys[0] {
```

Comment:
Can we assume that `hdr` has valid data if it's non-nil? It might be safer to add `err == nil`to this check.